ksql-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java

```diff
@@ -20,72 +20,116 @@
-    this.hasAggregateFunction = hasAggregateFunction;
+  void processGroupBy(final Expression expression) {
+    final AggregateVisitor visitor = new AggregateVisitor((aggFuncName, node) -> {
+      if (aggFuncName.isPresent()) {
+        throw new KsqlException("GROUP BY does not support aggregate functions: "
+            + aggFuncName.get() + " is an aggregate function.");
+      }
+      aggregateAnalysis.addGroupByField(node);
+    });
+
+    visitor.process(expression, new AnalysisContext());
   }
 
-  public AggregateAnalyzer(final AggregateAnalysis aggregateAnalysis,
-                           final Analysis analysis,
-                           final FunctionRegistry functionRegistry) {
-    this.aggregateAnalysis = aggregateAnalysis;
-    this.analysis = analysis;
-    this.functionRegistry = functionRegistry;
+  void processHaving(final Expression expression) {
+    final AggregateVisitor visitor = new AggregateVisitor((aggFuncName, node) -> {
+      if (!aggFuncName.isPresent()) {
+        aggregateAnalysis.addNonAggregateHavingField(node);
+      }
+    });
+    visitor.process(expression, new AnalysisContext());
   }
 
-  @Override
-  protected Node visitFunctionCall(final FunctionCall node, final AnalysisContext context) {
-    final String functionName = node.getName().getSuffix();
-    if (functionRegistry.isAggregate(functionName)) {
-      if (node.getArguments().isEmpty()) {
-        final Expression argExpression;
-        if (analysis.getJoin() != null) {
-          final Expression baseExpression = new QualifiedNameReference(
-              QualifiedName.of(analysis.getJoin().getLeftAlias()));
-          argExpression = new DereferenceExpression(baseExpression, SchemaUtil.ROWTIME_NAME);
-        } else {
-          final Expression baseExpression = new QualifiedNameReference(
-              QualifiedName.of(analysis.getFromDataSources().get(0).getRight()));
-          argExpression = new DereferenceExpression(baseExpression, SchemaUtil.ROWTIME_NAME);
+  private final class AggregateVisitor extends DefaultTraversalVisitor<Node, AnalysisContext> {
+
+    private final BiConsumer<Optional<String>, DereferenceExpression> dereferenceCollector;
+    private final Stack<String> aggregateFunctionStack = new Stack<>();
+    private boolean visitedAggFunction = false;
+
+    private AggregateVisitor(
+        final BiConsumer<Optional<String>, DereferenceExpression> dereferenceCollector
+    ) {
+      this.dereferenceCollector =
+          Objects.requireNonNull(dereferenceCollector, "dereferenceCollector");
+    }
+
+    @Override
+    protected Node visitFunctionCall(final FunctionCall node, final AnalysisContext context) {
+      final String functionName = node.getName().getSuffix();
+      final boolean aggregateFunc = functionRegistry.isAggregate(functionName);
+      if (aggregateFunc) {
+        visitedAggFunction = true;
+        if (node.getArguments().isEmpty()) {
+          node.getArguments().add(defaultArgument);
```

Comment:
Totally agree. But one for another PR.

Reply:
It's weird to me that we support this default at all. For example, we currently at this default to the aggregate functions `WindowStart()` and `WindowEnd()` which don't take any arguments (though it works fine since the arguments are just ignored), and we also blindly use ROWTIME as the default without checking whether the aggregate function accepts `BIGINT` or not (it so happens that all our one-argument aggregate functions today do).